server/auth/types/multiple/multi_auth.ts

```diff
@@ -0,0 +1,288 @@
+import { callTokenEndpoint } from '../openid/helper';
+import { MultiAuthRoutes } from './routes';
+import { SecuritySessionCookie } from '../../../session/security_cookie';
+
+export interface OpenIdAuthConfig {
+  authorizationEndpoint?: string;
+  tokenEndpoint?: string;
+  endSessionEndpoint?: string;
+  scope?: string;
+  authHeaderName?: string;
+}
+
+export interface WreckHttpsOptions {
+  ca?: string | Buffer | Array<string | Buffer>;
+  checkServerIdentity?: (host: string, cert: PeerCertificate) => Error | undefined;
+}
+
+export class MultipleAuthentication extends AuthenticationType {
+  private authTypes: string | string[];
+  private openIdAuthConfig: OpenIdAuthConfig;
+  private wreckClient: typeof wreck;
+
+  constructor(
+    config: SecurityPluginConfigType,
+    sessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>,
+    router: IRouter,
+    esClient: ILegacyClusterClient,
+    coreSetup: CoreSetup,
+    logger: Logger
+  ) {
+    super(config, sessionStorageFactory, router, esClient, coreSetup, logger);
+    this.authTypes = this.config.auth.type;
+    this.openIdAuthConfig = {};
+    this.wreckClient = this.createWreckClient();
+    this.init();
+  }
+
+  private async init() {
+    const routes = new MultiAuthRoutes(
+      this.router,
+      this.config,
+      this.sessionStorageFactory,
+      this.securityClient,
+      this.coreSetup
+    );
+
+    for (let i = 0; i < this.authTypes.length; i++) {
+      switch (this.authTypes[i].toLowerCase()) {
+        case AuthType.BASIC: {
+          routes.setupBasicRoutes();
+          break;
+        }
+        case AuthType.OPEN_ID: {
+          await this.oidcConfigSetup(this.wreckClient);
+          routes.setupOidcRoutes(this.openIdAuthConfig, this.wreckClient);
+          break;
+        }
+        case AuthType.SAML: {
+          routes.setupSamlRoutes();
+          break;
+        }
+        default: {
+          throw new Error(`Unsupported authentication type: ${this.authTypes[i]}`);
+        }
+      }
+    }
+  }
+
+  // override functions inherited from AuthenticationType
+  requestIncludesAuthInfo(
+    request: OpenSearchDashboardsRequest<unknown, unknown, unknown, any>
+  ): boolean {
+    return request.headers.authorization ? true : false;
+  }
+
+  getAdditionalAuthHeader(request: OpenSearchDashboardsRequest<unknown, unknown, unknown, any>) {
+    // This function is not used by basicauth, oidc and saml authentication, implementation is needed for jwt and proxy authentication.
+    return {};
+  }
+
+  async getCookie(
+    request: OpenSearchDashboardsRequest,
+    authInfo: any
+  ): Promise<SecuritySessionCookie> {
+    const sessionStore = await this.sessionStorageFactory.asScoped(request).get();
+    const reqAuthType = sessionStore?.authType;
+
+    if (
+      reqAuthType === AuthType.BASIC &&
+      this.config.auth.anonymous_auth_enabled &&
+      authInfo.user_name === OPENDISTRO_SECURITY_ANONYMOUS
```

Comment:
why we need to ensure authType is basicauth for anonymous users?

Reply:
Addressed.